src/execution/mod.rs

```diff
@@ -111,42 +112,42 @@ impl<'b> RuntimeInstance<'b> {
         wasm.move_start_to(inst.code_expr);
 
         loop {
-            match wasm.read_u8().unwrap_validated() {
+            match wasm.read_instruction().unwrap_validated() {
                 // end
-                0x0B => {
+                [NOP, ..] | [END, ..] => {
                     break;
                 }
                 // local.get: [] -> [t]
-                0x20 => {
+                [LOCAL_GET, ..] => {
                     let local_idx = wasm.read_var_u32().unwrap_validated() as LocalIdx;
                     let local = locals.get(local_idx);
                     trace!("Instruction: local.get [] -> [{local:?}]");
                     stack.push_value(local.clone());
                 }
                 // local.set [t] -> []
-                0x21 => {
+                [LOCAL_SET, ..] => {
                     let local_idx = wasm.read_var_u32().unwrap_validated() as LocalIdx;
                     let local = locals.get_mut(local_idx);
                     let value = stack.pop_value(local.to_ty());
                     trace!("Instruction: local.set [{local:?}] -> []");
                     *local = value;
                 }
```

Comment:
Wherever an instruction is followed by arguments, we could take these directly:

```rust
[LOCAL_SET, local_idx, ..] => {
  let local = locals.get_mut(local_idx);
  let value = stack.pop_value(local.to_ty());
  trace!("Instruction: local.set [{local:?}] -> []");
  *local = value;
}
```

Reply:
I think this could work, we also have to make sure we increment the pc and modify the `current` of the `WasmReader` instance accordingly

Reply:
Ideally we could get rid of the current in WasmReader IMHO

Reply:
Moved from `current` being a `&[u8]` to having a `pc` being a `usize`

Reply:
Still an architectural design that I need your input on. For each argument we still need to do a call to `strip_bytes`, do you want me to continue on this path? (read directly from the slice and call `strip_bytes`) Or do you want me to keep the current implementation?

Reply:
I would just like to add a few arguments against reading the arguments of operations while matching like this: `[SOME_OPCODE, some_argument, ..] => { ... }`.

1. This only works for arguments of fixed size. In WASM, however, integers are encoded with variable length (see [here](https://webassembly.github.io/spec/core/binary/values.html#integers)). Unfortunately, as far as I know, there is no elegant solution for matching them directly in the slice.
2. By matching the arguments directly in the slice, we decouple reading the argument bytes from advancing the program counter, which in my opinion is not very _rusty_. I think it is more error-prone, and puts the additional responsibility on the developer to make sure that the program counter is always incremented by the correct amount.

Don't get me wrong: The `strip_bytes` method is far from the best solution.
Maybe the best way is to write a macro with `macro_rules!` for the whole `match` statement, which allows us to match variable-length arguments. This macro could then also increment the program counter accordingly.